In [1]:
import pandas as pd
import numpy as np
from sklearn.neural_network import MLPClassifier
from sklearn import svm
from sklearn import preprocessing #to transform the feature labels
from sklearn.feature_extraction import DictVectorizer #to turn categorial variables into numeric arrays
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score

In [2]:
train_set = pd.read_csv("lab_4_training.csv")
test_set = pd.read_csv("lab_4_test.csv")
train_set.head()

Unnamed: 0 gender  age    year eyecolor  height  miles  brothers  sisters  \
0        1303   male   20  second    green      73  210.0         0        1   
1          36   male   20   third    other      71   90.0         1        0   
2         489   male   22  fourth    hazel      75  200.0         0        1   
3        1415   male   19  second    brown      72   35.0         2        2   
4         616   male   22  fourth    hazel      71   15.0         2        1   

   computertime exercise  exercisehours  musiccds  playgames  watchtv  
0          10.0      Yes            5.0        50        1.0     15.0  
1          15.0      Yes            4.0        10        0.0      1.0  
2           1.0      Yes            2.0       150        1.0     10.0  
3          20.0      Yes            5.0       100        0.0      7.0  
4          10.0      Yes            7.0        10        0.0      5.0

### 1. Calculate a baseline accuracy measure using the majority class
### a. Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [3]:
len(train_set[train_set["gender"] == "female"])/len(train_set) * 100

53.77358490566038

In [4]:
len(train_set[train_set["gender"] == "male"])/len(train_set) * 100

46.22641509433962

The majority class is 'female'. If we always predicted this class in the traning set, the accuracy would be 53.77%

### b. If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [5]:
len(test_set[test_set["gender"] == "female"])/len(test_set) * 100

52.26130653266332

If we always predicted 'female', the accuracy in the test set would be 52.26%

### 2. Get started with Neural Networks
### a. Choose a NN implementation and specify which you choose. Be sure the implementation allows you to modify the number of hidden layers and hidden nodes per layer.

sklearn.neural_network.MLPClassifier

### b. Train a neural network with a single 10 node hidden layer. Only use the Height feature of the dataset to predict the Gender. You will have to change Gender to a 0 and 1 class. After training, use your trained model to predict the class using the height feature from the training set. What was the accuracy of this prediction?

In [6]:
le = preprocessing.LabelEncoder()
le.fit(["male","female"])
target = le.transform(train_set.gender)
train_set_ht_np = np.reshape(train_set.height, (1590,1))

/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/numpy/core/fromnumeric.py:224: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  return reshape(newshape, order=order)


In [7]:
clf = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf.fit(train_set_ht_np, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [8]:
clf.score(train_set_ht_np, target)

0.53773584905660377

The accuracy is 84.65%

### c. Take the trained model from question 2.b and use it to predict the test set. This can be accomplished by taking the trained model and giving it the Height feature values from the test set. What is the accuracy of this model on the test set?

In [9]:
test_set_ht_np = np.reshape(test_set.height, (398,1))
test_target = le.transform(test_set.gender)

/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/numpy/core/fromnumeric.py:224: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  return reshape(newshape, order=order)


In [10]:
clf.score(test_set_ht_np, test_target)

0.52261306532663321

The accuracy is 85.43%

### d. Neural Networks tend to prefer smaller, normalized feature values. Try taking the log of the height feature in both training and testing sets or use a Standard Scalar operation in SKlearn to centre and normalize the data between 0-1 for continuous values. Repeat question 2.c with the log version and the normalized and centered version of this feature

In [11]:
min_max_scaler = preprocessing.MinMaxScaler()
train_set_ht_scaler = min_max_scaler.fit_transform(train_set.height.reshape(1590,1))
test_set_ht_scaler = min_max_scaler.transform(test_set.height.reshape(398,1))

/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  from ipykernel import kernelapp as app
/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype int64 was converted to float64 by MinMaxScaler.
  warnings.warn(msg, _DataConversionWarning)
/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:3: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  app.launch_new_instance()


In [12]:
clf_scaler = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_scaler.fit(train_set_ht_scaler, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [13]:
clf_scaler.score(train_set_ht_scaler, target)

0.84654088050314469

In [14]:
clf_scaler.score(test_set_ht_scaler, test_target)

0.85427135678391963

Normalizing the values to be between 0 - 1 didnt lead to any imporvement in accuracy (train = 84.65%, test = 85.43%)

In [15]:
train_set_ht_log = np.log10(train_set.height.reshape(1590,1))
test_set_ht_log = np.log10(test_set.height.reshape(398,1))

/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:1: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  if __name__ == '__main__':
/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  from ipykernel import kernelapp as app


In [16]:
clf_log = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_log.fit(train_set_ht_log, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [17]:
clf_log.score(train_set_ht_log, target)

0.84654088050314469

In [18]:
clf_log.score(test_set_ht_log, test_target)

0.85427135678391963

Taking log of the height values didnt lead to any imporvement in accuracy (train = 84.65%, test = 85.43%)

### 3. Get started with Support Vector Machines
### a. Chosen a SVM implementation and specify which you choose. Be sure the implementation allows you to choose between linear and RBF kernels.

sklearn.svm.SVC

### b. Use the same dataset from 2.b using the linear kernel to find training set prediction accuracy

In [19]:
clf_svm_linear = svm.SVC(kernel='linear')
clf_svm_linear.fit(train_set_ht_np, target)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [20]:
clf_svm_linear.score(train_set_ht_np, target)

0.83396226415094343

The training set prediction accuracy is 83.40%

### c. Use the same dataset from 2.b using the linear kernel to find test set prediction accuracy

In [21]:
clf_svm_linear.score(test_set_ht_np, test_target)

0.83165829145728642

The test set prediction accuracy is 83.17%

### d. Use the same dataset from 2.b using the RBF kernel to find training set prediction accuracy

In [22]:
clf_svm_rbf = svm.SVC(kernel='rbf')
clf_svm_rbf.fit(train_set_ht_np, target)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [23]:
clf_svm_rbf.score(train_set_ht_np, target)

0.84654088050314469

The training set prediction accuracy is 84.65%

### e. Use the same dataset from 2.b using the RBF kernel to find test set prediction accuracy

In [24]:
clf_svm_rbf.score(test_set_ht_np, test_target)

0.85427135678391963

The test set prediction accuracy is 85.43%

### f. Use the same dataset from 2.d (log) using the RBF to find test set prediction accuracy

In [25]:
clf_svm_rbf_log = svm.SVC(kernel='rbf')
clf_svm_rbf_log.fit(train_set_ht_log, target)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [26]:
clf_svm_rbf_log.score(test_set_ht_log, test_target)

0.83165829145728642

The test set preduction accuracy using log of height feature is 83.17%

### g. Z-score is a normalization technique. It is the value of a feature minus the average value for that feature in the training set, divided by the standard deviation of that feature in the training set. Repeat question 3.f using Z-score and note if there is any difference in accuracy and comment on why there is a change or no change in accuracy

In [27]:
z_scaler = preprocessing.StandardScaler()
train_set_ht_z_scaler = z_scaler.fit_transform(train_set.height.reshape(1590,1))
test_set_ht_z_scaler = z_scaler.transform(test_set.height.reshape(398,1))

/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  from ipykernel import kernelapp as app
/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, _DataConversionWarning)
/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:3: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  app.launch_new_instance()


In [28]:
clf_svm_rbf_scalar = svm.SVC(kernel='rbf')
clf_svm_rbf_scalar.fit(train_set_ht_z_scaler, target)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [29]:
clf_svm_rbf_scalar.score(test_set_ht_z_scaler, test_target)

0.85427135678391963

The test set prediction accuracy using z scaled values is 85.43% This is the same prediction accuracy as when using the actual height values (85.43%)

### 4. The rest of features in this dataset barring a few are categorical. Neither ML method accepts categorical features, so transform year, eyecolor, exercise into a set of binary features, one feature per unique original feature value, and mark the binary feature as ‘1’ if the feature value matches the original value and ‘0’ otherwise. Using only these binary variable transformed features, train and predict the class of the test set.

In [30]:
vec = DictVectorizer()
train_set_exercise_eyec_year = vec.fit_transform(train_set[['year','eyecolor','exercise']].to_dict('records')).toarray()
test_set_exercise_eyec_year = vec.transform(test_set[['year','eyecolor','exercise']].to_dict('records')).toarray()
vec.feature_names_

['exercise=No',
 'exercise=Yes',
 'eyecolor=blue',
 'eyecolor=brown',
 'eyecolor=green',
 'eyecolor=hazel',
 'eyecolor=other',
 'year=first',
 'year=fourth',
 'year=other',
 'year=second',
 'year=third']

### a. What was your accuracy using Neural Network with a single 10 node hidden layer? During training, use a maximum number of iterations of 50.

In [31]:
clf_eey = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,), max_iter=50)
clf_eey.fit(train_set_exercise_eyec_year, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=50, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [32]:
clf_eey.score(test_set_exercise_eyec_year, test_target)

0.57788944723618085

The test set prediction accuracy is 59.30%

### b. What was your accuracy using a SVM with RBF kernel?

In [33]:
clf_svm_rbf_eey = svm.SVC(kernel='rbf')
clf_svm_rbf_eey.fit(train_set_exercise_eyec_year, target)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

In [34]:
clf_svm_rbf_eey.score(test_set_exercise_eyec_year, test_target)

0.585427135678392

The test set prediction accuracy is 58.54%

### 5. Using a NN, does height + eye color predict the test set class better by
### a. Keeping the original feature values

In [35]:
vec = DictVectorizer()
train_set_eyec = vec.fit_transform(train_set[['height','eyecolor']].to_dict('records')).toarray()
test_set_eyec = vec.fit_transform(test_set[['height','eyecolor']].to_dict('records')).toarray()
vec.feature_names_

['eyecolor=blue',
 'eyecolor=brown',
 'eyecolor=green',
 'eyecolor=hazel',
 'eyecolor=other',
 'height']

In [36]:
clf_eh = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_eh.fit(train_set_eyec, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [37]:
clf_eh.score(test_set_eyec, test_target)

0.85929648241206025

Yes. The test set prediction accuracy increases significantly to 85.93%

### b. taking the log of the original values

In [38]:
train_set_eyec_log = np.append(train_set_eyec[:,:5],np.reshape(np.log10(train_set_eyec[:,5]),(1590,1)),axis=1)
test_set_eyec_log = np.append(test_set_eyec[:,:5],np.reshape(np.log10(test_set_eyec[:,5]),(398,1)),axis=1)

In [39]:
clf_eh_log = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_eh_log.fit(train_set_eyec_log, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [40]:
clf_eh_log.score(test_set_eyec_log, test_target)

0.85929648241206025

The test set prediction accuracy as compared to normal height values remains the same (85.92%)

### c. taking the Z-score of the original values

In [41]:
z_scaler = preprocessing.StandardScaler()
train_set_eyec_z_scaler = np.append(train_set_eyec[:,:5], z_scaler.fit_transform(np.reshape(train_set_eyec[:,5],(1590,1))),axis=1)
test_set_eyec_z_scaler = np.append(test_set_eyec[:,:5], z_scaler.fit_transform(np.reshape(test_set_eyec[:,5],(398,1))),axis=1)


In [42]:
clf_eh_z_scaler = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_eh_z_scaler.fit(train_set_eyec_z_scaler, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [43]:
clf_eh_z_scaler.score(test_set_eyec_z_scaler, test_target)

0.8693467336683417

The test set prediction accuracy as compared to normal height values increases to 87.16%

### 6. Repeat question 5 for exercise hours + eye color

In [44]:
vec = DictVectorizer()
train_set_eyeceh = vec.fit_transform(train_set[['eyecolor', 'exercisehours']].to_dict('records')).toarray()
test_set_eyeceh = vec.fit_transform(test_set[['eyecolor', 'exercisehours']].to_dict('records')).toarray()
vec.feature_names_

['exercisehours',
 'eyecolor=blue',
 'eyecolor=brown',
 'eyecolor=green',
 'eyecolor=hazel',
 'eyecolor=other']

In [45]:
clf_ee = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_ee.fit(train_set_eyeceh, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [46]:
clf_eh.score(test_set_eyeceh, test_target)

0.52261306532663321

The test set prediction accuracy is 52.26%

In [47]:
train_set_eyeceh_log = np.append(train_set_eyeceh[:,1:],np.reshape(np.log10(train_set_eyec[:,0]),(1590,1)),axis=1)
test_set_eyeceh_log = np.append(test_set_eyeceh[:,1:],np.reshape(np.log10(test_set_eyec[:,0]),(398,1)),axis=1)
#Fix -inf values
train_set_eyeceh_log[train_set_eyeceh_log==-np.inf] = -999
test_set_eyeceh_log[test_set_eyeceh_log==-np.inf]= -999

/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:1: RuntimeWarning: divide by zero encountered in log10
  if __name__ == '__main__':
/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: RuntimeWarning: divide by zero encountered in log10
  from ipykernel import kernelapp as app


In [48]:
clf_eh_log = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_eh_log.fit(train_set_eyeceh_log, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [49]:
clf_eh_log.score(test_set_eyeceh_log, test_target)

0.52261306532663321

The test set prediction accuracy remains the same 52.26%

In [50]:
z_scaler = preprocessing.StandardScaler()
train_set_eyeceh_z_scaler = np.append(train_set_eyeceh[:,1:], z_scaler.fit_transform(np.reshape(train_set_eyeceh[:,0],(1590,1))),axis=1)
test_set_eyeceh_z_scaler = np.append(test_set_eyeceh[:,1:], z_scaler.fit_transform(np.reshape(test_set_eyeceh[:,0],(398,1))),axis=1)


In [51]:
clf_eh_z_scaler = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_eh_z_scaler.fit(train_set_eyeceh_z_scaler, target)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='lbfgs', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [52]:
clf_eh_z_scaler.score(test_set_eyeceh_z_scaler, test_target)

0.59798994974874375

The test set prediction accuracy increases to 59.30%

### 7. Combine the features from question 4, 5, and exercise hours from question 6 (using the best normalization feature set form questions 5 and 6)

In [53]:
train_set_exercise_eyec_year_ht_eh = np.concatenate((train_set_exercise_eyec_year,z_scaler.fit_transform(np.reshape(train_set.height,(1590,1))),z_scaler.fit_transform(np.reshape(train_set.exercisehours,(1590,1)))),axis=1) 
test_set_exercise_eyec_year_ht_eh = np.concatenate((test_set_exercise_eyec_year,z_scaler.fit_transform(np.reshape(test_set.height,(398,1))),z_scaler.fit_transform(np.reshape(test_set.exercisehours,(398,1)))),axis=1) 
#Fix -inf values
train_set_exercise_eyec_year_ht_eh[train_set_exercise_eyec_year_ht_eh==-np.inf] = -999
test_set_exercise_eyec_year_ht_eh[test_set_exercise_eyec_year_ht_eh==-np.inf]= -999

/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/numpy/core/fromnumeric.py:224: FutureWarning: reshape is deprecated and will raise in a subsequent release. Please use .values.reshape(...) instead
  return reshape(newshape, order=order)
/Users/rohitraghavan/anaconda/lib/python3.5/site-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, _DataConversionWarning)


### a. What was the NN accuracy on the test set using the single 10 node hidden layer?

In [54]:
clf_comb = MLPClassifier(solver='lbfgs', activation='logistic', hidden_layer_sizes=(10,))
clf_comb.fit(train_set_exercise_eyec_year_ht_eh, target)
clf_comb.score(test_set_exercise_eyec_year_ht_eh, test_target)

0.89447236180904521

The test set prediction accuracy is 88.69%

### b. What was the SVM accuracy on the test set the RBF kernel?

In [55]:
clf_comb2 = svm.SVC(kernel='rbf')
clf_comb2.fit(train_set_exercise_eyec_year_ht_eh, target)
clf_comb2.score(test_set_exercise_eyec_year_ht_eh, test_target)

0.86683417085427139

The test set prediction accuracy is 86.68%

### 8. Can you improve your test set prediction accuracy by 5% or more?
### a. What was your best accuracy on the test set using the training set as your holdout? (for this question the training set is used for training AND testing)

In [56]:
# C_range = np.logspace(-2, 10, 4)
# gamma_range = np.logspace(-9, 3, 4)
C_range = range(-5, 16)
gamma_range = range(-15, 4)
best_score = 0
best_C = 0
best_gamma = 0
best_model = svm.SVC(kernel='rbf')
for C in C_range:
    for gamma in gamma_range:
        clf_best1 = svm.SVC(kernel='rbf', C=2**C, gamma=2**gamma)
        clf_best1.fit(train_set_exercise_eyec_year_ht_eh, target)
        score = clf_best1.score(train_set_exercise_eyec_year_ht_eh, target)
        if best_score < score:
            best_score = score
            best_C = C
            best_gamma = gamma
            best_model = clf_best1
print(best_score, best_C, best_gamma)

0.939622641509 6 3


In [58]:
best_model.score(test_set_exercise_eyec_year_ht_eh, test_target)

0.9346733668341709

In [57]:
clf_best1 = svm.SVC(kernel='rbf', C=2**best_C, gamma=2**best_gamma)
clf_best1.fit(train_set_exercise_eyec_year_ht_eh, target)
clf_best1.score(test_set_exercise_eyec_year_ht_eh, test_target)

0.9346733668341709

The best accuracy on the training set was 93.47%. On testing on the test set, the accuracy was 93.47

### b. What was your best accuracy on the test set using a 70% train 30% holdout?

In [59]:
train_set_exercise_eyec_year_ht_eh 
train_split_train, train_split_holdout, target_train_split_train, target_train_split_holdout = train_test_split(train_set_exercise_eyec_year_ht_eh, target, test_size=0.30, random_state=42)
best_score = 0
best_C = 0
best_gamma = 0
best_model = svm.SVC(kernel='rbf')
for C in C_range:
    for gamma in gamma_range:
        clf_best2 = svm.SVC(kernel='rbf', C=2**C, gamma=2**gamma)
        clf_best2.fit(train_split_train, target_train_split_train)
        score = clf_best2.score(train_split_holdout, target_train_split_holdout)
        if best_score < score:
            best_score = score
            best_C = C
            best_gamma = gamma
            best_model = clf_best2
print(best_score, best_C, best_gamma)

0.82180293501 -1 -6


In [61]:
best_model.score(test_set_exercise_eyec_year_ht_eh, test_target)

0.85678391959798994

In [60]:
clf_best2 = svm.SVC(kernel='rbf', C=2**best_C, gamma=2**best_gamma)
clf_best2.fit(train_set_exercise_eyec_year_ht_eh, target)
clf_best2.score(test_set_exercise_eyec_year_ht_eh, test_target)

0.85678391959798994

The best accuracy on validation set was 82.18%. On testing on the test set, the accuracy was 85.68%

### c. Finally, what was your best accuracy on the test set using 5-fold cross validation?

In [62]:
best_score = 0
best_C = 0
best_gamma = 0
best_model = svm.SVC(kernel='rbf')
for C in C_range:
    for gamma in gamma_range:
        clf_best3 = svm.SVC(kernel='rbf', C=2**C, gamma=2**gamma)
        scores = cross_val_score(clf_best3, train_set_exercise_eyec_year_ht_eh, target, cv=5)
        score = scores.mean()
        if best_score < score:
            best_score = score
            best_C = C
            best_gamma = gamma
            best_model = clf_best3
print(best_score, best_C, best_gamma)

0.853459119497 0 -8


In [63]:
clf_best3 = svm.SVC(kernel='rbf', C=2**best_C, gamma=2**best_gamma)
clf_best3.fit(train_set_exercise_eyec_year_ht_eh, target)
clf_best3.score(test_set_exercise_eyec_year_ht_eh, test_target)

0.86180904522613067

The best accuracy using K-fold cross validation was 86.18%

In [65]:
predictor = clf_best3.predict(train_set_exercise_eyec_year_ht_eh)
pd.DataFrame(predictor).to_csv("cv_predictions.txt", index =False)

In [66]:
predictor_test = clf_best3.predict(test_set_exercise_eyec_year_ht_eh)
pd.DataFrame(predictor_test).to_csv("test_set_prediction.txt", index =False)